[FLINK-7520][network] let our Buffer class extend from Netty's buffer class

NicoK opened the pull request:
## What is the purpose of the change

With this PR, we extend our own `Buffer` class to extend from Netty's `ByteBuf` class so that we can avoid one buffer copy while transferring data through Netty but keep our `MemorySegment` logic, performance, and configuration.

## Brief change log

- extract the `Buffer` interface (common functions used by non-Netty code inside Flink) and a `NetworkBuffer` implementation (extending from `ByteBuf`, implementing `Buffer`)
- change `Buffer` interface to follow the (separated) reader and writer index logic that Netty has and replace the `#getSize()`, ' #setSize()` logic, i.e.
```
    +-------------------+----------------+----------------+
    | discardable bytes | readable bytes | writable bytes |
    +-------------------+----------------+----------------+
    |                   |                |                |
    0      <=      readerIndex  <=  writerIndex   <=   size
```
(currently, only the writer index is used and both reading and writing in Flink code is performed exclusively either on NIO buffer wrappers or the underlying `MemorySegment`s directly)
- add `getNioBuffer()` and `getNioBufferReadable()` for properly accessing underlying buffer regions
- since we inherit from `AbstractByteBuf`, only one thread should work with the `Buffer`'s (meta)data as modifications to the indices are not thread-safe - this is the usual case though
- add `NetworkBuffer#setAllocator()` which is necessary to set before giving a `NetworkBuffer` into Netty code (we do not rely on this allocator in our code!)

## Verifying this change

This change added tests and can be verified as follows:

- extended `BufferTest` by inheriting from Netty's `AbstractByteBufTest` (copied into our sources due to it not being available in a separate test jar) to verify our buffer implementation follows Netty's invariants
- existing (integration) tests such as `NettyMessageSerializationTest` and `PartitionRequestClientHandlerTest` for the changes in the use of the new APIs
- any other (integration) test that uses the network stack

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dataartisans/flink/2)
<!-- Reviewable:end -->

Commit:
[hotfix][tests] replace DiscardingRecycler with FreeingBufferRecycler

Commit:
[hotfix][tests] do not use a mocked BufferRecycler for unpooled memory segments

The mock will actually keep references to the segments instead of freeing them.

Commit:
[hotfix][tests] make SpillableSubpartitionTest use TestBufferFactory.createBuffer

(this simplifies the test setups)

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/BufferTest.java:
```diff
@@ -53,9 +52,8 @@ protected ByteBuf newBuffer(int length, int maxCapacity) {
 				.allocateUnpooledSegment(Math.min(maxCapacity, MAX_CAPACITY_UPPER_BOUND));
-		final BufferRecycler recycler = Mockito.mock(BufferRecycler.class);
 
-		NetworkBuffer buffer = new NetworkBuffer(segment, recycler);
```
> The mock will actually keep references to the segments instead of freeing them.

Hahaha, I love this. Definitely will use it on my mockito praising lunch talk.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/iomanager/SynchronousBufferFileReader.java:
```diff
@@ -54,13 +56,13 @@ public void readInto(Buffer buffer) throws IOException {
-
-			fileChannel.read(buffer.getNioBuffer());
+			fileChannel.read(buffer.getNioBuffer(0, size));
+			buffer.setWriterIndex(size);
```
Please extract common logic of this and `BufferReadRequest` from above to some buffer reader class:


```
public class BufferFileChannelReader {
	private final ByteBuffer header = ByteBuffer.allocateDirect(8);

    public int readBufferLength(FileChannel fileChannel) {
            header.clear();
			fileChannel.read(header);
			header.flip();

			return header.getInt();
    }
 
   public boolean readBufferFromFileChannel(ByteBuffer buffer, FileChannel fileChannel) {
            int size = readBufferLength(fileChannel);
			final boolean isBuffer = size == 1;

			if (size > buffer.getMemorySegment().size()) {
				throw new IllegalStateException("Buffer is too small for data: " + buffer.getMemorySegment().size() + " bytes available, but " + size + " needed. This is most likely due to an serialized event, which is larger than the buffer size.");
			}

			buffer.setSize(size);

			fileChannel.read(buffer.getNioBuffer());

			if (!isBuffer) {
				buffer.tagAsEvent();
			}

			return fileChannel.size() - fileChannel.position() == 0;
    }
}
```

There is also some code duplication with `org.apache.flink.runtime.io.disk.iomanager.FileSegmentReadRequest#read`. That could re-use `readBufferLength` method.

and `FileSegment` doesn't ever use `isBuffer` field, so by removing it we could remove one more code duplication in `org.apache.flink.runtime.io.disk.iomanager.FileSegmentReadRequest#read`: `final boolean isBuffer = header.getInt() == 1;`

Sorry, I know that you just happened to touch this code, but this copy pasting really itches my brain 🙄 (and Let's Make the World a Better Place)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -280,7 +280,7 @@ public void writeBuffer(Buffer buffer, int subpartitionIndex) throws IOException
 				// Update statistics
 				totalNumberOfBuffers++;
-				totalNumberOfBytes += buffer.getSize();
+				totalNumberOfBytes += buffer.getWriterIndex();
```
I have some concerns regarding some of those changes.

For example this one suggests that `buffer` here is writable, while it's not. `getSize()` was more in line with "sealed" "read-only" buffers. It seems like some unnecessary abstractions are leaking here. 

Another things that worries me, is the sheer amount of those renames, which doesn't seem to be helping with code readability, but are tied with underlying `Buffer` implementation. If we choose to refactor underlying `Buffer` next time, will it force us again to rename all of those things?

Maybe we should wrap/proxy some methods, that are meaningful in our context (like `ReadOnlyBuffer.getSize()`) into `NettyBuffer` specifc implementation (like it just happen to be `NettyBuffer.getWriterIndex()`)?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -195,7 +197,7 @@ BufferAndAvailability getNextBuffer() throws IOException {
 		}
 
-		numBytesIn.inc(next.getSizeUnsafe());
+		numBytesIn.inc(next.getWriterIndex());
```
same here. `getWriterIndex()` is not exact logical replacement of the `getSizeUnsafe()`. It's more like in our case those two methods will return the same thing, but the underlying meaning is a different.

I think we should keep using `Buffer.getSize()` method in most of those places.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/iomanager/BufferFileWriterReaderTest.java:
```diff
@@ -199,25 +198,26 @@ private int getNextMultipleOf(int number, int multiple) {
-	public static int fillBufferWithAscendingNumbers(Buffer buffer, int currentNumber) {
-		MemorySegment segment = buffer.getMemorySegment();
+	static int fillBufferWithAscendingNumbers(Buffer buffer, int currentNumber, int size) {
+		assert size % 4 == 0;
```
`checkState`?

NicoK replied:
`checkArgument` - but why not (although iirc, assertions should be enabled in tests)

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/AbstractByteBufTest.java:
```diff
@@ -0,0 +1,3077 @@
+ *
+ * Copied from netty 4.0.50 with some changes to fit our netty version 4.0.27.
+ */
+public abstract class AbstractByteBufTest {
```
What's the purpose of those tests? Aren't they testing `AbstractReferenceCountedByteBuf` code which is tested on it's in netty?

NicoK replied:
Actually, the test itself is abstract for any `ByteBuf` implementation and it does verify ours.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java:
```diff
@@ -42,6 +42,7 @@ public static Buffer createMockBuffer(int size) {
 		final Buffer mockBuffer = mock(Buffer.class);
 		when(mockBuffer.isBuffer()).thenReturn(true);
 		when(mockBuffer.getSize()).thenReturn(size);
+		when(mockBuffer.getWriterIndex()).thenReturn(size);
```
:( Please create a real `Buffer` here

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionTest.java:
```diff
@@ -193,7 +194,7 @@ public void notifyBuffersAvailable(long numBuffers) {
 		SpillableSubpartition partition = createSubpartition();
 
-		Buffer buffer = new Buffer(MemorySegmentFactory.allocateUnpooledSegment(4096), FreeingBufferRecycler.INSTANCE);
+		Buffer buffer = new NetworkBuffer(MemorySegmentFactory.allocateUnpooledSegment(4096), FreeingBufferRecycler.INSTANCE, true, 4096);
```
`TestBufferFactory.createBuffer(4096);`

probably same could be applied to many some places?

NicoK replied:
actually, that's what the included `[hotfix][tests] make SpillableSubpartitionTest use TestBufferFactory.createBuffer` commit does...

NicoK replied:
oh - a classical race condition here - I fixed this one day before your review and put this onto the branch ;)

pnowojski reviewed (changes requested):
I don't have any critical objections about the general idea, especially if it lets us avoid additional data copy :)

Commit:
[hotfix][tests] replace InputChannelTestUtils#createMockBuffer() with TestBufferFactory#createBuffer()

This eliminates one more unnecessary buffer mock.

Commit:
[hotfix][io] remove duplicate code between SynchronousBufferFileReader and BufferReadRequest

Commit:
[FLINK-7520][network] let our Buffer class extend from netty's buffer class

For this, use a common (flink) Buffer interface and an implementation
(NetworkBuffer) that implements netty's buffer methods as well. In the future,
with this, we are able to avoid unnecessary buffer copies when handing buffers
over to netty while keeping our MemorySegment logic and configuration.

For the netty-specific part, the NetworkBuffer also requires a ByteBuf allocator
which is otherwise not needed in our use cases, so if the buffer is handed over
to netty, it requires a byte buffer allocator to be set.

Commit:
[hotfix][network] clarify BufferResponse#size() uses (by removing it)

This field was only used by the code paths on the receiver and was inconsistent
with what was added on the sending side. We should use the contained buffer's
readableBytes() instead, depending on the actual use case.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/Buffer.java:
```diff
@@ -22,141 +22,153 @@
+	 * @return writer index (from 0 (inclusive) to the size of the backing {@link MemorySegment}
+	 * (inclusive))
+	 */
+	int getSizeUnsafe();
```
nit: now I have some doubts about `getSizeUnsafe()`. We should drop it and replace by `getSize()` itself, but maybe that could/should be done in separate PR?

NicoK replied:
I kept it for now

pnowojski reviewed (approved):
LGTM

one nit comment which you can ignore or fix (whatever seems like a better solution in your opinion)

NicoK commented:
rebased and squashed - creating a PR towards apache/flink now
